### ISSUE TITLE
PEImage:Load(void* buffer, size_t) - does not work ?

### ISSUE THREAD

**Issue by MINIONBOTS** [DarthTon/Blackbone#99]
Sorry to open 2 issues a day, but I really want to get my stuff running again in x64 ;)

Before, in 32bit, I was reading my dll into memory and passing a pointer of it directly to _process.mmap().MapImage(...), this worked fine all the time.
Now I am trying it in x64 and this did not work anymore. So I looked into Xenos's code and adapted the usage of PEImage:Load().

This works:
```
NTSTATUS status = img.Load(L"C:\\mydll.dll");
...
_process.mmap().MapImage(img.imageSize(), img.base(), true, flags)

```

This does not work:
```

HANDLE  _hFile = CreateFileW(
		L"C:\\mydll.dll", FILE_GENERIC_READ,
		FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
		NULL, OPEN_EXISTING, 0, NULL
	);

	if (_hFile != INVALID_HANDLE_VALUE)
	{
               void* _pFileBase = 0;
		bool _isPlainData = false;
		HANDLE _hMapping = CreateFileMappingW(_hFile, NULL, SEC_IMAGE | PAGE_READONLY, 0, 0, NULL);
		if (_hMapping && _hMapping != INVALID_HANDLE_VALUE)	{
			_isPlainData = false;
			_pFileBase = MapViewOfFile(_hMapping, FILE_MAP_READ, 0, 0, 0);
		}
		else{
			_isPlainData = true;
			_hMapping = CreateFileMappingW(_hFile, NULL, PAGE_READONLY, 0, 0, NULL);

			if (_hMapping && _hMapping != INVALID_HANDLE_VALUE)
				_pFileBase = MapViewOfFile(_hMapping, FILE_MAP_READ, 0, 0, 0);
		}
                blackbone::pe::PEImage img;
		NTSTATUS status = img.Load(_pFileBase, 6652928);  
...
_process.mmap().MapImage(img.imageSize(), img.base(), true, flags)
...
```

^This is basicly exactly what blackbone's PEImage:Load(wstring path) is doing internally, and it does not work with the PEImage:Load(buffer,flags) method....which is strange. 
I ofcourse tried a "normal" way of reading the file into memory, using stringstream and .str().c_str() to get a continous buffer pointer for the Load() function, the result is the same. 

(I am still running on /MT if that matters...have to spend a bit more time on the weekend to get MD of all libs I use , but I updated with your latest commit...not sure if that matters at all)

Long story short, could maybe anyone confirm that this PEImage:Load(buffer, size) method works in x64 manual mapping ?  And why is the "size" argument being passed but never used in the function itself ? 

Thanks & good night ^^

**Comment by DarthTon** [comment]
Does 'TestMMapFromMem' sample work in your case? It does work for your test dll on my PC
```C++
HANDLE hFile = CreateFileW( L"C:\\!Testing\\Win32Project1.dll", FILE_GENERIC_READ, 0x7, 0, OPEN_EXISTING, 0, 0 );
    if (hFile != INVALID_HANDLE_VALUE)
    {
        DWORD bytes = 0;
        size = GetFileSize( hFile, NULL );
        buf = VirtualAlloc( NULL, size, MEM_COMMIT, PAGE_READWRITE );
        ReadFile( hFile, buf, size, &bytes, NULL );
        CloseHandle( hFile );
    }

    if (thisProc.mmap().MapImage( size, buf, false, NoFlags ) == 0)
    {
        std::wcout << L"Mapping failed with error 0x" << std::hex << LastNtStatus()
                   << L". " << Utils::GetErrorDescription( LastNtStatus() ) << std::endl << std::endl;
    }
    else
        std::wcout << L"Successfully mapped, unmapping\n";
```

**Comment by MINIONBOTS** [comment]
Thanks for the answer, I'm currently not able to test it, but on Wednesday I'll give that a shot and reply back.

**Comment by manuel76413** [comment]
I meet the same problem with MINIONBOTS  also. I have a 3rd library which only have a dll and a head file, I cannot modify the function in the dll. For this dll, it is succeed to use LoadLibrary, and the GetProcAddress, and then call function ok.

Now I need to Load it from memory for some reason, I selected YOUR BlackBone Library to do this.
However, I tried many many times,  still cannot do this OK, my code is like :

	std::wstring dllpath1 = L"E:\\e.dll";
	PEImage img;
	NTSTATUS status = img.Load(dllpath1.c_str());

	NTSTATUS status1 = Driver().DisableDEP(GetCurrentProcessId());  //I try to supress SEH exception
	if (0==thisProc.mmap().MapImage(img.imageSize(), img.base(), false, NoFlags) ) {
		std::wcout << L"Mapping failed with error 0x" << std::hex << LastNtStatus()
		           << L". " << Utils::GetErrorDescription(LastNtStatus()) << std::endl << std::endl;
	} else {
		std::wcout << L"Successfully mapped, unmapping\n";		
		SAFE_CALL(Open);
       }

I don't known if my code is correct, however it raised error in manual load mode.

attached is my test dll, can you test it for me, THANKS!

**Comment by DarthTon** [comment]
```C++
thisProc.mmap().MapImage(img.imageSize(), img.base(), **true**, NoFlags)
```

**Comment by manuel76413** [comment]
no true or false I test both, this 3rd dll looks like compiled with delphi
So I googled long time, many coder say , cannot load from memory for SEH exception.

I find I cannot upload attach for some unknown reasons,
do you have email ?

**Comment by DarthTon** [comment]
Yes - darthton@gmail.com

**Comment by manuel76413** [comment]
Hi DarthTon , did you receive my mail ?

**Comment by DarthTon** [comment]
Yeap, it went to spam for reason though.

**Comment by manuel76413** [comment]
Hi DarthTon, Thanks for your quickly reply.  I'm very glad to hear you solved this.
Could you paste sample code here, Thanks.

**Comment by DarthTon** [comment]
```C++
HANDLE hFile = CreateFileW( L"C:\\\HqL1.dll", FILE_GENERIC_READ, 0x7, 0, OPEN_EXISTING, 0, 0 );
if (hFile != INVALID_HANDLE_VALUE)
{
    DWORD bytes = 0;
    size = GetFileSize( hFile, NULL );
    buf = VirtualAlloc( NULL, size, MEM_COMMIT, PAGE_READWRITE );
    ReadFile( hFile, buf, size, &bytes, NULL );
    CloseHandle( hFile );
}

auto pImage = thisProc.mmap().MapImage( size, buf, false, CreateLdrRef );

VirtualFree( buf, 0, MEM_RELEASE );
auto pfn = thisProc.modules().GetExport( pImage, "TdxHq_Connect" );

char Result[1024] = { 0 };
char ErrInfo[256] = { 0 };
RemoteFunction< int( __stdcall*)(char*, int, char*, char*)> rfn( thisProc, pfn.procAddress, (char*)"180.153.18.171", 7709, (char*)Result, (char*)ErrInfo );
rfn.setArg( 2, AsmVariant( Result, 1024 ) );
rfn.setArg( 3, AsmVariant( ErrInfo, 256 ) );
int res = 0;
rfn.Call( res );
```

**Comment by manuel76413** [comment]
It is OK now.  you are so  wonderful, and the blackbone you created is amazing!
though the function call method is too difficult for me. I think you can simplify this or write more sample code or document.

Thank you, master!

**Comment by MINIONBOTS** [comment]
I finally had time to look at it again. I load my dll data into memory with std::ostringstream ostream;

This does NOT work:

```
size_t size = ostream.str().length();
PEImage img;
img.Load(ostream.str().c_str() ,size)
thisProc.mmap().MapImage( img.imageSize(), img.base(), true, NoFlags );
```


And this works:
```
size_t size = ostream.str().length();
thisProc.mmap().MapImage( size, ostream.str().c_str(), false, NoFlags );
```

I don't know why it does not work when using the PEImage class, (yes I tried changing the "plaindata" and "asimage" bool, same result)

Long  story short, after nearly 2,5days starring at it, I guess I'll just skip the PEIMage class and go with the working version. Maybe you have another idea what I could try, but if not, then just close this issue.

THanks a lot for you work! (Think about the patreon link :) )

**Comment by DarthTon** [comment]
What's 'ostream' in your case. Can you post full code?

**Comment by DarthTon** [comment]
PEImage::Load from memory doesn't modify image mapping type, i.e. if file was loaded as simple data file it won't be modified to image type. Only loading from file can create data in image format.
So in your case:
```C++
size_t size = ostream.str().length();
PEImage img;
img.Load(ostream.str().c_str() ,size)
thisProc.mmap().MapImage( img.imageSize(), img.base(), false, NoFlags );
```
